### PR TITLE
core: use TryGetAccount to read what TryUpdateAccount has written

### PR DESCRIPTION
**Pull request by gballet**
This is a PR with a smaller footprint than #25431 . It introduces `TryGetAccount`, a function that will return a `types.StateAccount` without making any assumption as to what format the account is encoded with in the database.

### Rationale

The encoding of the leaf in RLP is very much dependent on the format of the tree, and performing the encoding in the state database is causing an abstraction leak: Verkle trees no not use RLP, and therefore a distinction on the type of tree has to be made at the database level, which imo should be abstracted to the tree itself.

Instead of writing the following in `statedb.go`'s `getDeletedStateObject`:

```golang
// If snapshot unavailable or reading from it failed, load from the database
	if data == nil {
        	var enc []byte
        	var err error
        	data = new(types.StateAccount)

		start := time.Now()
		// if this is verkle, read each account element one by one...
		if s.trie.IsVerkle() {
			data.version =  s.trie.TryGet(addr.Bytes, 0)
			data.Balance =  s.trie.TryGet(addr.Bytes, 1)
			data.Nonce =  s.trie.TryGet(addr.Bytes, 2)
			data.CodeHash =  s.trie.TryGet(addr.Bytes, 4)
		} else {
			// if this is the MPT, read the RLP
			enc, err = s.trie.TryGet(addr.Bytes())
		}
		if metrics.EnabledExpensive {
			s.AccountReads += time.Since(start)
		}
		if err != nil {
			s.setError(fmt.Errorf("getDeleteStateObject (%x) error: %v", addr.Bytes(), err))
			return nil
		}
		if !s.trie.IsVerkle() {
			if len(enc) == 0 {
				return nil
			}
		
			if err := rlp.DecodeBytes(enc, data); err != nil {
				log.Error("Failed to decode state object", "addr", addr, "err", err)
				return nil
			}
		}
	}
	// Insert into the live set
	obj := newObject(s, addr, *data)
	s.setStateObject(obj)
	return obj
}
```

one can simply write:

```golang
if data == nil {
		start := time.Now()
		var err error
		data, err = s.trie.TryGetAccount(addr.Bytes())
		if metrics.EnabledExpensive {
			s.AccountReads += time.Since(start)
		}
		if err != nil {
			s.setError(fmt.Errorf("getDeleteStateObject (%x) error: %w", addr.Bytes(), err))
			return nil
		}
		if data == nil {
			return nil
		}
	}
```

i.e. the exact format of the account encoding is abstracted away to the trie's `TryGet` function and there is no longer a need to call `IsVerkle()` on the tree.

### CODE REVIEW

File: trie/trie.go [hunk 1]
```diff
@@ -295,7 +312,9 @@ func (t *Trie) TryUpdateAccount(key []byte, acc *types.StateAccount) error {
 	if err != nil {
 		return fmt.Errorf("can't encode object at %x: %w", key[:], err)
 	}
-	return t.TryUpdate(key, data)
+	// Encoding []byte cannot fail, ok to ignore the error.
+	v, _ := rlp.EncodeToBytes(common.TrimLeftZeroes(data[:]))
```
Comment (rjl493456442) [inline trie/trie.go:316]: Is it a mistake? You encode twice.

File: core/state/database.go [hunk 2]
```diff
@@ -71,7 +71,9 @@ type Trie interface {
 	// trie.MissingNodeError is returned.
 	TryGet(key []byte) ([]byte, error)
 
-	// TryUpdateAccount abstract an account write in the trie.
+	TryGetAccount(key []byte) (*types.StateAccount, error)
```
Comment (rjl493456442) [inline core/state/database.go:74]: Please add the function descrpiton.

File: eth/protocols/snap/handler.go [hunk 3]
```diff
@@ -419,8 +418,9 @@ func ServiceGetStorageRangesQuery(chain *core.BlockChain, req *GetStorageRangesP
 			if err != nil {
 				return nil, nil
 			}
-			var acc types.StateAccount
-			if err := rlp.DecodeBytes(accTrie.Get(account[:]), &acc); err != nil {
+			var acc *types.StateAccount
```
Comment (rjl493456442) [inline eth/protocols/snap/handler.go:421]: We can simplify it to 

```go
acc, err := accTrie.TryGetAccount(account[:])
if err != nil {
	return nil, nil
}
```

It's totally fine to redefine the error.

File: trie/secure_trie.go [hunk 4]
```diff
@@ -96,7 +100,7 @@ func (t *SecureTrie) TryUpdateAccount(key []byte, acc *types.StateAccount) error
 	if err != nil {
 		return err
 	}
-	if err := t.trie.TryUpdate(hk, data); err != nil {
+	if err := t.trie.tryUpdate(hk, data); err != nil {
```
Comment (rjl493456442) [inline trie/secure_trie.go:103]: I feel like `TryUpdateAccount` and `TryGetAccount` are not aligned in SecureTrie.
Personally I think we should move all account encode/decode logic to Trie and SecureTrie is just a wrapper for hashing key.

File: trie/trie.go [hunk 5]
```diff
@@ -154,6 +154,23 @@ func (t *Trie) Get(key []byte) []byte {
 	return res
 }
 
+func (t *Trie) TryGetAccount(key []byte) (*types.StateAccount, error) {
+	res, newroot, didResolve, err := t.tryGet(t.root, keybytesToHex(key), 0)
+	if err != nil {
+		log.Error(fmt.Sprintf("Unhandled trie error: %v", err))
+		return &types.StateAccount{}, err
+	}
+	if didResolve {
```
Comment (rjl493456442) [inline trie/trie.go:163]: Why not use `t.TryGet`? It will set the root for you.

**Comment by rjl493456442** [comment]
I am wondering do we need to provide a specific API for retrieving storage slot as well? How do we retrieve storage in verkle?

**Comment by gballet** [comment]
> I am wondering do we need to provide a specific API for retrieving storage slot as well? How do we retrieve storage in verkle?

we do need, but this is waaaay more complicated and the subject of a different PR that is still work in progress

File: trie/trie.go (hunk 1 above)
Comment (gballet) [reply trie/trie.go]: yup, it's weird that it even worked :thinking:

**Review by holiman — APPROVED** [review]
LGTM

**Comment by gballet** [comment]
This lasts commit implement a suggestion by Gary to demote `trie.Trie` from a structure implementing `state.Trie` into a simple backend, and to leave that role to `trie.SecureTrie`. In this model, `trie.Trie` becomes a backend store for `trie.SecureTrie` and is no longer concerned with RLP encoding or the hashing of keys.

Verkle trees already follow this model, with `VerkleTrie` that takes care of transforming accounts into several leaves, and also turning an account address and a slot address into a tree path. The tree management itself is left to `go-verkle`, that will be merged into this repo in the long term.

`SecureTrie` is also renamed `StateTrie`.

File: eth/protocols/snap/handler.go [hunk 6]
```diff
@@ -415,15 +413,15 @@ func ServiceGetStorageRangesQuery(chain *core.BlockChain, req *GetStorageRangesP
...
 			// Request started at a non-zero hash or was capped prematurely, add
 			// the endpoint Merkle proofs
-			accTrie, err := trie.New(common.Hash{}, req.Root, chain.StateCache().TrieDB())
+			accTrie, err := trie.NewSecure(common.Hash{}, req.Root, chain.StateCache().TrieDB())
 			if err != nil {
 				return nil, nil
 			}
-			var acc types.StateAccount
-			if err := rlp.DecodeBytes(accTrie.Get(account[:]), &acc); err != nil {
+			acc, err := accTrie.TryGetAccount(account[:])
```
Comment (rjl493456442) [inline eth/protocols/snap/handler.go:420]: Hmm, it's wrong here. The provided `account` is already the hash of the account address.

Comment (rjl493456442) [reply eth/protocols/snap/handler.go:420]: Maybe we can add one more function in `trie.StateTrie` via which we can retrieve account with accountHash directly.

File: trie/secure_trie.go [hunk 7]
```diff
@@ -54,20 +58,20 @@ type SecureTrie struct {
 // Loaded nodes are kept around until their 'cache generation' expires.
 // A new cache generation is created by each call to Commit.
 // cachelimit sets the number of past cache generations to keep.
-func NewSecure(owner common.Hash, root common.Hash, db *Database) (*SecureTrie, error) {
+func NewSecure(owner common.Hash, root common.Hash, db *Database) (*StateTrie, error) {
```
Comment (rjl493456442) [inline trie/secure_trie.go:61]: Perhaps also rename it to `NewStateTrie`?

**Review by rjl493456442 — APPROVED** [review]
lgtm

**Comment by fjl** [comment]
Commit prefix should've been `core/state: ...` because that's where the interface change is.

File: trie/secure_trie.go [hunk 8]
```diff
@@ -78,19 +88,50 @@ func (t *SecureTrie) Get(key []byte) []byte {
...
+func (t *StateTrie) TryGet(key []byte) ([]byte, error) {
 	return t.trie.TryGet(t.hashKey(key))
 }
 
+func (t *StateTrie) TryGetAccount(key []byte) (*types.StateAccount, error) {
+	var ret types.StateAccount
+	res, err := t.trie.TryGet(t.hashKey(key))
+	if err != nil {
+		log.Error(fmt.Sprintf("Unhandled trie error: %v", err))
+		return &ret, err
```
Comment (holiman) [inline trie/secure_trie.go:100]: Feels dangerous to return a non-nil empty `types.StateAccount` here

File: trie/secure_trie.go [hunk 9]
```diff
@@ -78,19 +88,50 @@ func (t *SecureTrie) Get(key []byte) []byte {
...
+	}
+	if res == nil {
+		return nil, nil
+	}
+	err = rlp.DecodeBytes(res, &ret)
+	return &ret, err
```
Comment (holiman) [inline trie/secure_trie.go:106]: And here

File: trie/secure_trie.go [hunk 10]
```diff
@@ -78,19 +88,50 @@ func (t *SecureTrie) Get(key []byte) []byte {
...
+
+// TryGetAccountWithPreHashedKey does the same thing as TryGetAccount, however
+// it expects a key that is already hashed. This constitutes an abstraction leak,
+// since the client code needs to know the key format.
+func (t *StateTrie) TryGetAccountWithPreHashedKey(key []byte) (*types.StateAccount, error) {
+	var ret types.StateAccount
+	res, err := t.trie.TryGet(key)
+	if err != nil {
+		log.Error(fmt.Sprintf("Unhandled trie error: %v", err))
+		return &ret, err
```
Comment (holiman) [inline trie/secure_trie.go:117]: And here

**Pull request merged**